libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp

```diff
@@ -0,0 +1,331 @@
+#include <utility>
+
+#include <GCS_MAVLink/GCS.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define BATTMONITOR_XRAY_NUM_CELLS      3
+#define BATTMONITOR_ENDURANCE_NUM_CELLS 6
+
+#define BATTMONITOR_SMBUS_SUI_BUTTON_DEBOUNCE      3       // button held down for 3 intervals will cause a power off event
+
+#define REG_CELL_VOLTAGE        0x28
+#define REG_CURRENT             0x2a
+
+const int AVG_SIZE = 10;
+
+static float sum(std::vector<float>& v) {
+    float val = 0.0f;
+
+    for(std::vector<float>::iterator e = v.begin(); e != v.end(); e++) {
+        val += *e;
+    }
+
+    return val;
+}
+
+static float avg(std::vector<float>& values) {
+    return sum(values) / values.size();
+}
+
+/*
+ * Other potentially useful registers, listed here for future use
+ * #define BATTMONITOR_SMBUS_SUI_VOLTAGE           0x09    // voltage register
+ * #define BATTMONITOR_SMBUS_SUI_BATTERY_STATUS    0x16    // battery status register including alarms
+ * #define BATTMONITOR_SMBUS_SUI_DESIGN_CAPACITY   0x18    // design capacity register
+ * #define BATTMONITOR_SMBUS_SUI_DESIGN_VOLTAGE    0x19    // design voltage register
+ * #define BATTMONITOR_SMBUS_SUI_SERIALNUM         0x1c    // serial number register
+ * #define BATTMONITOR_SMBUS_SUI_MANUFACTURE_NAME  0x20    // manufacturer name
+ * #define BATTMONITOR_SMBUS_SUI_DEVICE_NAME       0x21    // device name
+ * #define BATTMONITOR_SMBUS_SUI_DEVICE_CHEMISTRY  0x22    // device chemistry
+ * #define BATTMONITOR_SMBUS_SUI_MANUFACTURE_INFO  0x25    // manufacturer info including cell voltage
+ */
+
+// Constructor
+AP_BattMonitor_SMBus_SUI::AP_BattMonitor_SMBus_SUI(AP_BattMonitor &mon,
+                    AP_BattMonitor::BattMonitor_State &mon_state,
+                    AP_BattMonitor_Params &params,
+                    AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev,
+                    uint8_t cell_count)
+    : AP_BattMonitor_SMBus(mon, mon_state, params, std::move(dev)),
+     _cell_count(MIN(4, cell_count)),
+    _capacity(0)
+{
+    _pec_supported = false;
+    _dev->set_retries(2);
+}
+
+void AP_BattMonitor_SMBus_SUI::init(void)
+{
+    AP_BattMonitor_SMBus::init();
+    if (_dev && timer_handle) {
+        // run twice as fast for two phases
+        _dev->adjust_periodic_callback(timer_handle, 50000);
+    }
+}
+
+void AP_BattMonitor_SMBus_SUI::timer()
+{
+    uint8_t buff[8];
+    uint32_t tnow = AP_HAL::micros();
+
+    // we read in two phases as the device can stall if you read
+    // current too rapidly after voltages
+    phase_voltages = !phase_voltages;
+
+    if (phase_voltages) {
+        read_cell_voltages();
+        return;
+    }
+
+    // read current
+    if (read_block_bare(REG_CURRENT, buff, 4, false)) {
+        float current_amps = -(float)((int32_t)((uint32_t)buff[3]<<24 | (uint32_t)buff[2]<<16 | (uint32_t)buff[1]<<8 | (uint32_t)buff[0])) / 1000.0f;
+
+        currents.push_back(current_amps);
+        if(currents.size() > 3) {
+            currents.erase(currents.begin());
```

Comment:
Doesn't this make the average divide by 0 right after this?

Reply:
fixed